src/main/kotlin/com/openlattice/shuttle/IntegrationService.kt

```diff
@@ -0,0 +1,349 @@
+        integration.maxConnections.ifPresent { dstDataSource.maximumPoolSize = it }
+
+        val pgDestination = PostgresDestination(
+                entitySets.mapKeys { it.value.id },
+                entityTypes,
+                propertyTypes.mapKeys { it.value.id },
+                dstDataSource
+        )
+
+        if (s3BucketUrl.isBlank()) {
+            return mapOf(StorageDestination.POSTGRES to pgDestination)
+        }
+        val s3Api = Retrofit.Builder()
+                .baseUrl(s3BucketUrl)
+                .addConverterFactory(RhizomeByteConverterFactory())
+                .addConverterFactory(RhizomeJacksonConverterFactory(ObjectMappers.getJsonMapper()))
+                .addCallAdapterFactory(RhizomeCallAdapterFactory())
+                .client(RetrofitBuilders.okHttpClient().build())
+                .build().create(S3Api::class.java)
+        val s3Destination = PostgresS3Destination(pgDestination, s3Api, dataIntegrationApi)
+        return mapOf(StorageDestination.POSTGRES to pgDestination, StorageDestination.S3 to s3Destination)
+    }
+
+    private fun createLogEntitySet(integrationName: String, integration: Integration): UUID {
+        val name = buildLogEntitySetName(integrationName)
+        val contacts = integration.contacts
+        val description = buildLogEntitySetDescription(integration.flightPlanParameters)
+        val logEntitySet = EntitySet(
+                logEntityType.id,
+                name,
+                name,
+                Optional.of(description),
+                contacts,
+                Optional.empty(),
+                IdConstants.GLOBAL_ORGANIZATION_ID.id,
+                Optional.empty(),
+                Optional.empty()
+        )
+        return entitySetManager.createEntitySet(Principals.getCurrentUser(), logEntitySet)
+    }
+
+    private fun buildLogEntitySetName(integrationName: String): String {
+        var name = "Integration logs for $integrationName"
+        var nameAttempt = name
+        var count = 1
+
+        while (reservationService.isReserved(nameAttempt)) {
+            nameAttempt = "${name}_$count"
+            count++
+        }
+
+        return nameAttempt
+    }
+
+    private fun buildLogEntitySetDescription(flightPlanParameters: Map<String, FlightPlanParameters>): String {
+        val flightDescriptionsClause = flightPlanParameters.values.joinToString(", ") {
+            val flight = it.flight!!
+            val tags = if (flight.tags.isEmpty()) {
+                ""
+            } else {
+                " with tags [${flight.tags.joinToString(", ")}]"
+            }
+            return@joinToString "${flight.name}$tags"
+        }
+
+        return "Auto-generated entity set containing logs of the following flights: $flightDescriptionsClause"
+    }
+
+    private fun generateIntegrationJobId(): UUID {
+        var jobId = UUID.randomUUID()
+        while (integrationJobs.containsKey(jobId)) {
```

Comment:
this is HIGHLY unlikely to happen, but you could theoretically run into a race condition here where two integrations get the same jobId, both are validated, and then they both write to the map. The safer thing would be to take the IntegrationJob as a parameter here, and then do 
```
while (integrationJobs.putIfAbsent(jobId, integrationJob) != null) {
    jobId = UUID.randomUUID() 
}```